In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("./abstract_metadata_with_ideal_spans_correct_no_spans_with_min4chars_df_en_cz_final_with_selected_8_docs.tsv", sep = '\t')

In [3]:
first_4_text_list = (3, 7, 6, 19)

In [4]:
filtered_df_first_4 = df[df['hal_id'].isin(first_4_text_list)]


In [5]:
filtered_df_first_4


,hal_id,translation_id,sentence_indices,num_sentences,total_characters,source,total_spans,total_span_chars,span_char_ratio_pct
0,3,3,"[0, 1, 2, 3, 4, 5, 6, 7]",8,1785,Using data and artificial intelligence to try ...,42,600,33.61
1,7,7,"[24, 25, 26, 27, 28, 29, 30, 31]",8,1625,LNG investments hit record in 2019 & the bigge...,29,401,24.68
6,6,6,"[16, 17, 18, 19, 20, 21, 22, 23]",8,1367,Three Scottish students named among Europe's b...,30,574,41.99
7,19,19,"[56, 57, 58, 59, 60, 61, 62, 63]",8,1356,"US sends troops, air defense systems to Saudi ...",35,457,33.70


In [6]:
values = filtered_df_first_4[['hal_id','total_spans']]
values

,hal_id,total_spans
0,3,42
1,7,29
6,6,30
7,19,35


In [8]:
import json

In [9]:
path_json = "spans_ideal_df_en_cz_final_with_selected_8_docs.json"
# Load JSON data
with open(path_json, 'r', encoding='utf-8') as f:
    json_data_ideal_spans = json.load(f)

In [10]:
import pandas as pd
path_csv = "postedition_aligned_with_tokenized_offsets_df_en_cz_final_with_selected_8_docs.csv"
df_en_cz = pd.read_csv(path_csv)

**Get sentence length for each abstract**

In [11]:
def sentence_length_abstract(translations):
    sentences_length_of_abstract = []
    for idx, value in translations.items():
        sentences_length_of_abstract.append(len(value))
    return sentences_length_of_abstract


In [14]:
sentences_length_of_abstract = sentence_length_abstract(df_en_cz.iloc[0:8]['translation'])


In [15]:
sentences_length_of_abstract

[227, 257, 159, 236, 246, 182, 245, 233]

In [17]:
print(sum(sentences_length_of_abstract))

1785


In [93]:
import ast

In [137]:
spans = json_data_ideal_spans[1009:1017]
#spans = json_data_ideal_spans[894:901]


In [138]:
spans

[[{'text': ['la'], 'start': 46, 'end': 49},
  {'text': ['du', 'dialogue', 'abstrait'], 'start': 67, 'end': 88}],
 [{'text': ['pré-formés'], 'start': 25, 'end': 36},
  {'text': ['du', 'langage', 'naturel,'], 'start': 97, 'end': 117}],
 [{'text': ['du', 'langage'], 'start': 103, 'end': 114},
  {'text': ['la', 'préformation'], 'start': 188, 'end': 204}],
 [],
 [{'text': ['pré-formés'], 'start': 131, 'end': 142},
  {'text': ['langue:'], 'start': 159, 'end': 167},
  {'text': ['pré-formés', 'multilingues:'], 'start': 211, 'end': 236}],
 [{'text': ['synopses', 'abstraites'], 'start': 113, 'end': 133},
  {'text': ['agents'], 'start': 216, 'end': 223}],
 [{'text': ['la', 'technologie'], 'start': 110, 'end': 125}],
 [{'text': ['pré-formés'], 'start': 50, 'end': 61}]]

In [139]:
spans[1][0]

{'text': ['pré-formés'], 'start': 25, 'end': 36}

In [140]:
spans[1][0]['start']

25

In [141]:
# Define the slice range
start_index = 894
end_index = 901
slice_data = spans
print(slice_data)
# Collect absolute indices of sentences that will be empty after filtering
sentence_indices_no_spans = []
sentence_indices_with_spans = []
for index, spans in enumerate(spans):
    no_spans = 0
    print("spans: ", spans)
    print(index)
    number_of_spans = 0
    if spans:

        for span in spans:
            if span['end'] - span['start'] >= 4:
                number_of_spans += 1
                continue
            else:
                no_spans += 1
            number_of_spans += 1
            print("number_of_spans: ", number_of_spans)
    else:
        print("i am at else")
        sentence_indices_no_spans.append(index)
    print("no_spans: ", no_spans, ", number_of_spans: ", number_of_spans)
    if no_spans == number_of_spans and number_of_spans!=0:
        sentence_indices_no_spans.append(index)
   
    sentence_indices_with_spans.append(index)


# Output the result
print("Indices of sentences with no spans after filtering:", sentence_indices_no_spans)
print("Indices of sentences with no spans after filtering:", sentence_indices_with_spans)

[[{'text': ['la'], 'start': 46, 'end': 49}, {'text': ['du', 'dialogue', 'abstrait'], 'start': 67, 'end': 88}], [{'text': ['pré-formés'], 'start': 25, 'end': 36}, {'text': ['du', 'langage', 'naturel,'], 'start': 97, 'end': 117}], [{'text': ['du', 'langage'], 'start': 103, 'end': 114}, {'text': ['la', 'préformation'], 'start': 188, 'end': 204}], [], [{'text': ['pré-formés'], 'start': 131, 'end': 142}, {'text': ['langue:'], 'start': 159, 'end': 167}, {'text': ['pré-formés', 'multilingues:'], 'start': 211, 'end': 236}], [{'text': ['synopses', 'abstraites'], 'start': 113, 'end': 133}, {'text': ['agents'], 'start': 216, 'end': 223}], [{'text': ['la', 'technologie'], 'start': 110, 'end': 125}], [{'text': ['pré-formés'], 'start': 50, 'end': 61}]]
spans:  [{'text': ['la'], 'start': 46, 'end': 49}, {'text': ['du', 'dialogue', 'abstrait'], 'start': 67, 'end': 88}]
0
number_of_spans:  1
no_spans:  1 , number_of_spans:  2
spans:  [{'text': ['pré-formés'], 'start': 25, 'end': 36}, {'text': ['du', 'l

In [143]:
# Remove all elements in list1 from list2
sentence_indices_with_spans = [x for x in sentence_indices_with_spans if x not in sentence_indices_no_spans]

print(sentence_indices_with_spans)


[0, 1, 2, 4, 5, 6, 7]


In [142]:
list1 = [3,7]
list2 = [0, 1, 2, 3, 4, 5, 6, 7]

# Remove all elements in list1 from list2
list2 = [x for x in list2 if x not in list1]

print(list2)


[0, 1, 2, 4, 5, 6]


In [70]:
type(spans)

list

In [71]:
valid_sentence_indexes = []


In [17]:
spans = [[{'text': ['la'], 'start': 46, 'end': 49},
  {'text': ['du', 'dialogue', 'abstrait'], 'start': 67, 'end': 88}],
 [{'text': ['pré-formés'], 'start': 25, 'end': 36},
  {'text': ['du', 'langage', 'naturel,'], 'start': 97, 'end': 117}],
 [{'text': ['du', 'langage'], 'start': 103, 'end': 114},
  {'text': ['la', 'préformation'], 'start': 188, 'end': 204}],
 [],
 [{'text': ['pré-formés'], 'start': 131, 'end': 142},
  {'text': ['langue:'], 'start': 159, 'end': 167},
  {'text': ['pré-formés', 'multilingues:'], 'start': 211, 'end': 236}],
 [{'text': ['synopses', 'abstraites'], 'start': 113, 'end': 133},
  {'text': ['agents'], 'start': 216, 'end': 223}],
 [{'text': ['la', 'technologie'], 'start': 110, 'end': 125}],
 [{'text': ['pré-formés'], 'start': 50, 'end': 61}]]

In [18]:
spans_with_new_starts_ends = []
previous_sentence_length = 2

In [19]:
for index, spans in enumerate(spans):
    if index == 0:
        spans_title = []
        for span in spans:
            if span['start'] == 0:
                start = 0
            else:
                start = span['start'] + 1
            spans_title.append({
                                'text': span['text'][:],
                                'startnew': start,
                                'endnew': span['end']
                                
                                })

        spans_with_new_starts_ends.append(spans_title)

    # elif index == 1:
    #     #sentences
    #     spans_each_sentence = []
    #     for span in spans:
    #         spans_each_sentence.append({
    #                             'text': span['text'],
    #                             'startnew': span['start'] + 1 +2,
    #                             'endnew': span['end']
                                
    #                             })

    #     spans_with_new_starts_ends.append(spans_each_sentence)
    elif index == 1:
        spans_each_sentence = []
        previous_sentence_length += sentences_length_of_abstract[index - 1]
        print("index is: ", index, " and previous sentence lenght: ", previous_sentence_length)
        
        for ind,span in enumerate(spans):
            print("index in index 1: ", ind)
            print("spans in index 1: ", span)
            if span['start'] == 0:
                start = span['start'] + previous_sentence_length
            else:
                start = span['start'] + previous_sentence_length + 1
           
            spans_each_sentence.append({
                                'text': span['text'][:],
                                'startnew': start,
                                'endnew': span['end'] + previous_sentence_length
                                
                                })
        spans_with_new_starts_ends.append(spans_each_sentence)

    else:
        spans_each_sentence = []
        previous_sentence_length += sentences_length_of_abstract[index - 1] + 1
        print("index is: ", index, " and previous sentence lenght: ", previous_sentence_length)
        for ind,span in enumerate(spans):
            if span['start'] == 0:
                start = span['start'] + previous_sentence_length
            else:
                start = span['start'] + previous_sentence_length + 1
            spans_each_sentence.append({
                                'text': span['text'][:],
                                'startnew': start,
                                'endnew': span['end'] + previous_sentence_length
                                
                                })

        spans_with_new_starts_ends.append(spans_each_sentence)

index is:  1  and previous sentence lenght:  90
index in index 1:  0
spans in index 1:  {'text': ['pré-formés'], 'start': 25, 'end': 36}
index in index 1:  1
spans in index 1:  {'text': ['du', 'langage', 'naturel,'], 'start': 97, 'end': 117}
index is:  2  and previous sentence lenght:  401
index is:  3  and previous sentence lenght:  633
index is:  4  and previous sentence lenght:  724
index is:  5  and previous sentence lenght:  985
index is:  6  and previous sentence lenght:  1238
index is:  7  and previous sentence lenght:  1386


In [20]:
spans_with_new_starts_ends

[[{'text': ['la'], 'startnew': 47, 'endnew': 49},
  {'text': ['du', 'dialogue', 'abstrait'], 'startnew': 68, 'endnew': 88}],
 [{'text': ['pré-formés'], 'startnew': 116, 'endnew': 126},
  {'text': ['du', 'langage', 'naturel,'], 'startnew': 188, 'endnew': 207}],
 [{'text': ['du', 'langage'], 'startnew': 505, 'endnew': 515},
  {'text': ['la', 'préformation'], 'startnew': 590, 'endnew': 605}],
 [],
 [{'text': ['pré-formés'], 'startnew': 856, 'endnew': 866},
  {'text': ['langue:'], 'startnew': 884, 'endnew': 891},
  {'text': ['pré-formés', 'multilingues:'], 'startnew': 936, 'endnew': 960}],
 [{'text': ['synopses', 'abstraites'], 'startnew': 1099, 'endnew': 1118},
  {'text': ['agents'], 'startnew': 1202, 'endnew': 1208}],
 [{'text': ['la', 'technologie'], 'startnew': 1349, 'endnew': 1363}],
 [{'text': ['pré-formés'], 'startnew': 1437, 'endnew': 1447}]]

In [21]:
for index, spans in enumerate(spans_with_new_starts_ends):
    if spans:
        for span in spans:
            print("texts in this span actual: ", span["text"])
            print("locate in this text: ", text935[span['startnew']: span['endnew']])
        print("============new sentence==========")

texts in this span actual:  ['la']
locate in this text:  la
texts in this span actual:  ['du', 'dialogue', 'abstrait']
locate in this text:  du dialogue abstrait
============new sentence==========
texts in this span actual:  ['pré-formés']
locate in this text:  pré-formés
texts in this span actual:  ['du', 'langage', 'naturel,']
locate in this text:  du langage naturel,
============new sentence==========
texts in this span actual:  ['du', 'langage']
locate in this text:  du langage
texts in this span actual:  ['la', 'préformation']
locate in this text:  la préformation
============new sentence==========
texts in this span actual:  ['pré-formés']
locate in this text:  pré-formés
texts in this span actual:  ['langue:']
locate in this text:  langue:
texts in this span actual:  ['pré-formés', 'multilingues:']
locate in this text:  pré-formés multilingues:
============new sentence==========
texts in this span actual:  ['synopses', 'abstraites']
locate in this text:  synopses abstraites
text

In [22]:
import pandas as pd
path_csv = "/home/arafat/vscode/NN/postedition_aligned_with_tokenized_offsets_translator.csv"
df_translators_word_offsets = pd.read_csv(path_csv)

In [23]:
import ast

# Convert the string representations to actual Python objects (list of tuples)
word_offsets_935 = df_translators_word_offsets.iloc[1009:1017]["translation_word_offset"].apply(ast.literal_eval)

# Now iterate through the converted values
for idx, value in word_offsets_935.items():
    print(f"Index: {idx}, Value: {value}")  # value is now a list of tuples


Index: 1009, Value: [(0, 10), (10, 14), (14, 22), (22, 25), (25, 32), (32, 42), (42, 46), (46, 49), (49, 55), (55, 58), (58, 67), (67, 70), (70, 79), (79, 88)]
Index: 1010, Value: [(0, 3), (3, 11), (11, 25), (25, 36), (36, 40), (40, 47), (47, 54), (54, 57), (57, 63), (63, 67), (67, 76), (76, 83), (83, 86), (86, 97), (97, 100), (100, 108), (108, 117), (117, 119), (119, 127), (127, 130), (130, 139), (139, 142), (142, 152), (152, 156), (156, 163), (163, 166), (166, 174), (174, 184), (184, 195), (195, 199), (199, 212), (212, 217), (217, 227), (227, 230), (230, 238), (238, 252), (252, 257), (257, 260), (260, 270), (270, 283), (283, 286), (286, 295), (295, 310)]
Index: 1011, Value: [(0, 10), (10, 13), (13, 18), (18, 28), (28, 33), (33, 40), (40, 51), (51, 53), (53, 59), (59, 64), (64, 68), (68, 76), (76, 84), (84, 88), (88, 91), (91, 103), (103, 106), (106, 114), (114, 123), (123, 127), (127, 139), (139, 148), (148, 158), (158, 163), (163, 167), (167, 174), (174, 183), (183, 188), (188, 191)

In [24]:
word_offsets_935

1009    [(0, 10), (10, 14), (14, 22), (22, 25), (25, 3...
1010    [(0, 3), (3, 11), (11, 25), (25, 36), (36, 40)...
1011    [(0, 10), (10, 13), (13, 18), (18, 28), (28, 3...
1012    [(0, 2), (2, 9), (9, 12), (12, 19), (19, 28), ...
1013    [(0, 4), (4, 7), (7, 16), (16, 21), (21, 32), ...
1014    [(0, 3), (3, 15), (15, 19), (19, 23), (23, 33)...
1015    [(0, 3), (3, 13), (13, 22), (22, 26), (26, 30)...
1016    [(0, 4), (4, 14), (14, 17), (17, 23), (23, 27)...
Name: translation_word_offset, dtype: object

In [25]:
counter = 0
updated_word_offsets_935 = []
previous_sentence_length = 2

In [26]:
sentences_length_of_abstract

[88, 310, 231, 90, 260, 252, 147, 122]

In [27]:
for index, value in word_offsets_935.items():
    if counter == 0:
        updated_offsets_each_sentence = []
        for offset in value:
            if offset[0] == 0:
                start = 0
            else:
                start = offset[0] + 1
            updated_offsets_each_sentence.append((start, offset[1]))

        updated_word_offsets_935.append(updated_offsets_each_sentence)

    elif counter == 1:
        updated_offsets_each_sentence = []
        previous_sentence_length += sentences_length_of_abstract[counter - 1]
        for ind,offset in enumerate(value):
            if ind == 0:
                start = offset[0]+previous_sentence_length
            else:
                start = offset[0]+previous_sentence_length + 1
            updated_offsets_each_sentence.append((start, offset[1]+previous_sentence_length))

        updated_word_offsets_935.append(updated_offsets_each_sentence)
    else:
        updated_offsets_each_sentence = []
        previous_sentence_length += sentences_length_of_abstract[counter - 1] + 1
        for ind,offset in enumerate(value):
            if ind == 0:
                start = offset[0]+previous_sentence_length
            else:
                start = offset[0]+previous_sentence_length + 1
            updated_offsets_each_sentence.append((start, offset[1]+previous_sentence_length))
        updated_word_offsets_935.append(updated_offsets_each_sentence)
    counter += 1


In [28]:
updated_word_offsets_935

[[(0, 10),
  (11, 14),
  (15, 22),
  (23, 25),
  (26, 32),
  (33, 42),
  (43, 46),
  (47, 49),
  (50, 55),
  (56, 58),
  (59, 67),
  (68, 70),
  (71, 79),
  (80, 88)],
 [(90, 93),
  (94, 101),
  (102, 115),
  (116, 126),
  (127, 130),
  (131, 137),
  (138, 144),
  (145, 147),
  (148, 153),
  (154, 157),
  (158, 166),
  (167, 173),
  (174, 176),
  (177, 187),
  (188, 190),
  (191, 198),
  (199, 207),
  (208, 209),
  (210, 217),
  (218, 220),
  (221, 229),
  (230, 232),
  (233, 242),
  (243, 246),
  (247, 253),
  (254, 256),
  (257, 264),
  (265, 274),
  (275, 285),
  (286, 289),
  (290, 302),
  (303, 307),
  (308, 317),
  (318, 320),
  (321, 328),
  (329, 342),
  (343, 347),
  (348, 350),
  (351, 360),
  (361, 373),
  (374, 376),
  (377, 385),
  (386, 400)],
 [(401, 411),
  (412, 414),
  (415, 419),
  (420, 429),
  (430, 434),
  (435, 441),
  (442, 452),
  (453, 454),
  (455, 460),
  (461, 465),
  (466, 469),
  (470, 477),
  (478, 485),
  (486, 489),
  (490, 492),
  (493, 504),
  (505, 

In [145]:
updated_word_offsets_935[0]

[(0, 10),
 (11, 14),
 (15, 22),
 (23, 25),
 (26, 32),
 (33, 42),
 (43, 46),
 (47, 49),
 (50, 55),
 (56, 58),
 (59, 67),
 (68, 70),
 (71, 79),
 (80, 88)]

In [29]:
for idx, value in enumerate(updated_word_offsets_935):
    print(f"Index: {idx}, Value: {value}")
    for offset in value:
        #print(offset)
        print(text935[offset[0]:offset[1]])

Index: 0, Value: [(0, 10), (11, 14), (15, 22), (23, 25), (26, 32), (33, 42), (43, 46), (47, 49), (50, 55), (56, 58), (59, 67), (68, 70), (71, 79), (80, 88)]
Efficacité
des
modèles
de
langue
française
sur
la
tâche
de
synthèse
du
dialogue
abstrait
Index: 1, Value: [(90, 93), (94, 101), (102, 115), (116, 126), (127, 130), (131, 137), (138, 144), (145, 147), (148, 153), (154, 157), (158, 166), (167, 173), (174, 176), (177, 187), (188, 190), (191, 198), (199, 207), (208, 209), (210, 217), (218, 220), (221, 229), (230, 232), (233, 242), (243, 246), (247, 253), (254, 256), (257, 264), (265, 274), (275, 285), (286, 289), (290, 302), (303, 307), (308, 317), (318, 320), (321, 328), (329, 342), (343, 347), (348, 350), (351, 360), (361, 373), (374, 376), (377, 385), (386, 400)]
Les
modèles
linguistiques
pré-formés
ont
établi
l’état
de
l’art
sur
diverses
tâches
de
traitement
du
langage
naturel,
y
compris
la
synthèse
du
dialogue,
qui
permet
au
lecteur
d’accéder
rapidement
aux
informations
clés
prove

Correction et validation de la dépendance syntaxique dans le Rhapsodie de la banque d’arbres française parlée

Cet article présente les méthodes, les résultats et la précision du processus d’annotation syntaxique de la Rhapsodie Treebank de français parlé. Le Rhapsodie Treebank est un corpus de 33 000 mots annoté pour la prosodie et la syntaxe, sous licence dans son intégralité sous Creative Commons. L’annotation syntaxique contient deux niveaux: un niveau macro-syntactique, contenant une segmentation en unités illocutionnaires (y compris des marqueurs de discours, des parenthèses...) et un niveau micro-syntactique comprenant des relations de dépendance et diverses structures paradigmatiques, appelées constructions de pile, ces dernières étant particulièrement fréquentes et diversifiées dans le langage parlé. Le processus d’annotation micro-syntactique, présenté dans cet article, comprend une préparation semi-automatique de la transcription, l’application d’un analyseur de dépendance syntaxique, le transcodage des résultats d’analyse au schéma d’annotation Rhapsodie, la correction manuelle par plusieurs annotateurs suivie d’un processus de validation, et enfin l’application de règles de cohérence qui vérifient les erreurs courantes. Les bons scores d’accord inter-annotateurs sont présentés et analysés plus en détail. L’article comprend également la liste des fonctions utilisées dans l’annotation de dépendance et pour la distinction des différentes constructions de pieux et présente les idées qui sous-tendent ces choix.

[{'text': 'Correction et validation de la dépendance syntaxique dans le Rhapsodie de la banque d’arbres française parlée\n\n',
  'start': 0,
  'end': 111},
 {'text': 'Cet article présente les méthodes, les résultats et la précision du processus d’',
  'start': 111,
  'end': 191},
 {'text': 'annotation syntaxique de la Rhapsodie Treebank de français parlé.',
  'start': 191,
  'end': 256},
 {'text': 'Le Rhapsodie Treebank est un corpus de 33\xa0000 mots annoté pour la prosodie et la syntaxe, sous licence dans son intégralité sous Creative Commons.',
  'start': 257,
  'end': 403},
 {'text': 'L’annotation syntaxique contient deux niveaux: un niveau macro-syntactique, contenant une segmentation en unités illocutionnaires (y compris des marqueurs de discours, des parenthèses...) et un niveau micro-syntactique comprenant des relations de dépendance et diverses structures paradigmatiques, appelées constructions de pile, ces dernières étant particulièrement fréquentes et diversifiées dans le langage parlé.',
  'start': 404,
  'end': 820},
 {'text': 'Le processus d’annotation micro-syntactique, présenté dans cet article, comprend une préparation semi-automatique de la transcription, l’application d’un analyseur de dépendance syntaxique, le transcodage des résultats d’analyse au schéma d’annotation Rhapsodie, la correction manuelle par plusieurs annotateurs suivie d’un processus de validation, et enfin l’application de règles de cohérence qui vérifient les erreurs courantes.',
  'start': 821,
  'end': 1252},
 {'text': 'Les bons scores d’accord inter-annotateurs sont présentés et analysés plus en détail.',
  'start': 1253,
  'end': 1338},
 {'text': 'L’article comprend également la liste des fonctions utilisées dans l’annotation de dépendance et pour la distinction des différentes constructions de pieux et présente les idées qui sous-tendent ces choix.',
  'start': 1339,
  'end': 1544}]

In [30]:


# Filtered result
filtered_spans_min_4_char = [
    [span for span in group if span['endnew'] - span['startnew'] >= 4]
    for group in spans_with_new_starts_ends
]

# Optional: remove empty groups
filtered_spans_min_4_char = [group for group in filtered_spans_min_4_char if group]

# Print or use the result
print(filtered_spans_min_4_char)


[[{'text': ['du', 'dialogue', 'abstrait'], 'startnew': 68, 'endnew': 88}], [{'text': ['pré-formés'], 'startnew': 116, 'endnew': 126}, {'text': ['du', 'langage', 'naturel,'], 'startnew': 188, 'endnew': 207}], [{'text': ['du', 'langage'], 'startnew': 505, 'endnew': 515}, {'text': ['la', 'préformation'], 'startnew': 590, 'endnew': 605}], [{'text': ['pré-formés'], 'startnew': 856, 'endnew': 866}, {'text': ['langue:'], 'startnew': 884, 'endnew': 891}, {'text': ['pré-formés', 'multilingues:'], 'startnew': 936, 'endnew': 960}], [{'text': ['synopses', 'abstraites'], 'startnew': 1099, 'endnew': 1118}, {'text': ['agents'], 'startnew': 1202, 'endnew': 1208}], [{'text': ['la', 'technologie'], 'startnew': 1349, 'endnew': 1363}], [{'text': ['pré-formés'], 'startnew': 1437, 'endnew': 1447}]]


In [31]:
filtered_spans_min_4_char

[[{'text': ['du', 'dialogue', 'abstrait'], 'startnew': 68, 'endnew': 88}],
 [{'text': ['pré-formés'], 'startnew': 116, 'endnew': 126},
  {'text': ['du', 'langage', 'naturel,'], 'startnew': 188, 'endnew': 207}],
 [{'text': ['du', 'langage'], 'startnew': 505, 'endnew': 515},
  {'text': ['la', 'préformation'], 'startnew': 590, 'endnew': 605}],
 [{'text': ['pré-formés'], 'startnew': 856, 'endnew': 866},
  {'text': ['langue:'], 'startnew': 884, 'endnew': 891},
  {'text': ['pré-formés', 'multilingues:'], 'startnew': 936, 'endnew': 960}],
 [{'text': ['synopses', 'abstraites'], 'startnew': 1099, 'endnew': 1118},
  {'text': ['agents'], 'startnew': 1202, 'endnew': 1208}],
 [{'text': ['la', 'technologie'], 'startnew': 1349, 'endnew': 1363}],
 [{'text': ['pré-formés'], 'startnew': 1437, 'endnew': 1447}]]

In [101]:
for index, spans in enumerate(filtered_spans_min_4_char):
    print(index)

0
1
2
3
4
5
6


In [32]:
noised_spans_min_4_char = []

In [33]:
import random

def create_shifted_annotations(spans, offsets):
    shifted_annotations = []
    
    for i, sentence_spans in enumerate(spans):
        sent_offsets = offsets[i]
        new_sentence_annotations = []
        
        for span_dict in sentence_spans:
            orig_start = span_dict['startnew']
            orig_end = span_dict['endnew']
            
            start_token_idx = None
            end_token_idx = None
            
            for idx, (s, e) in enumerate(sent_offsets):
                if s <= orig_start < e:
                    start_token_idx = idx
                if s < orig_end <= e:
                    end_token_idx = idx
            
            if start_token_idx is None or end_token_idx is None:
                start_token_idx = 0
                end_token_idx = len(sent_offsets) - 1
            
            annotations_for_this_span = []
            for _ in range(2):
                which_end = random.choice(['start', 'end'])
                direction = random.choice(['L', 'R'])
                step = random.choice([1, 2, 3, 4])
                
                new_start_idx = start_token_idx
                new_end_idx = end_token_idx
                
                if which_end == 'start':
                    if direction == 'L':
                        new_start_idx = max(0, start_token_idx - step)
                    else:
                        new_start_idx = min(len(sent_offsets) - 1, start_token_idx + step)
                    if new_start_idx > end_token_idx:
                        new_start_idx = end_token_idx
                else:
                    if direction == 'L':
                        new_end_idx = max(0, end_token_idx - step)
                    else:
                        new_end_idx = min(len(sent_offsets) - 1, end_token_idx + step)
                    if new_end_idx < start_token_idx:
                        new_end_idx = start_token_idx
                
                new_start_char = sent_offsets[new_start_idx][0]
                new_end_char = sent_offsets[new_end_idx][1]
                
                if new_start_char < orig_end and orig_start < new_end_char:
                    if which_end == 'start':
                        if direction == 'L':
                            alt_start_idx = min(len(sent_offsets) - 1, start_token_idx + step)
                            if alt_start_idx > end_token_idx:
                                alt_start_idx = end_token_idx
                        else:
                            alt_start_idx = max(0, start_token_idx - step)
                        alt_start_char = sent_offsets[alt_start_idx][0]
                        alt_end_char = sent_offsets[end_token_idx][1]
                        if alt_start_char < orig_end and orig_start < alt_end_char:
                            final_start_char = sent_offsets[start_token_idx][0]
                            final_end_char = sent_offsets[end_token_idx][1]
                        else:
                            final_start_char = alt_start_char
                            final_end_char = alt_end_char
                    else:
                        if direction == 'L':
                            alt_end_idx = min(len(sent_offsets) - 1, end_token_idx + step)
                            if alt_end_idx < start_token_idx:
                                alt_end_idx = start_token_idx
                        else:
                            alt_end_idx = max(0, end_token_idx - step)
                        alt_start_char = sent_offsets[start_token_idx][0]
                        alt_end_char = sent_offsets[alt_end_idx][1]
                        if alt_start_char < orig_end and orig_start < alt_end_char:
                            final_start_char = sent_offsets[start_token_idx][0]
                            final_end_char = sent_offsets[end_token_idx][1]
                        else:
                            final_start_char = alt_start_char
                            final_end_char = alt_end_char
                else:
                    final_start_char = new_start_char
                    final_end_char = new_end_char
                
                annotations_for_this_span.append({
                    'startnew': final_start_char,
                    'endnew': final_end_char
                })
            
            new_sentence_annotations.append(annotations_for_this_span)
        
        shifted_annotations.append(new_sentence_annotations)
    
    return shifted_annotations

# Example usage:

shifted_annots = create_shifted_annotations(filtered_spans_min_4_char, updated_word_offsets_935)

In [34]:
shifted_annots

[[[{'startnew': 68, 'endnew': 67}, {'startnew': 68, 'endnew': 88}]],
 [[{'startnew': 116, 'endnew': 126}, {'startnew': 116, 'endnew': 126}],
  [{'startnew': 188, 'endnew': 176}, {'startnew': 188, 'endnew': 207}]],
 [[{'startnew': 505, 'endnew': 515}, {'startnew': 505, 'endnew': 515}],
  [{'startnew': 590, 'endnew': 605}, {'startnew': 590, 'endnew': 605}]],
 [[{'startnew': 646, 'endnew': 723}, {'startnew': 633, 'endnew': 708}],
  [{'startnew': 633, 'endnew': 723}, {'startnew': 633, 'endnew': 723}],
  [{'startnew': 633, 'endnew': 723}, {'startnew': 653, 'endnew': 723}]],
 [[{'startnew': 724, 'endnew': 984}, {'startnew': 724, 'endnew': 960}],
  [{'startnew': 724, 'endnew': 976}, {'startnew': 724, 'endnew': 984}]],
 [[{'startnew': 1001, 'endnew': 1237}, {'startnew': 985, 'endnew': 1237}]],
 [[{'startnew': 1238, 'endnew': 1385}, {'startnew': 1238, 'endnew': 1385}]]]

In [35]:
sentences_length_of_abstract

[88, 310, 231, 90, 260, 252, 147, 122]

In [36]:
sentences_length_of_abstract_indexes = []
for index, sentence in enumerate(updated_word_offsets_935):
    
    sentences_length_of_abstract_indexes.append((sentence[0][0], sentence[-1][1]))
sentences_length_of_abstract_indexes

[(0, 88),
 (90, 400),
 (401, 632),
 (633, 723),
 (724, 984),
 (985, 1237),
 (1238, 1385),
 (1386, 1508)]

In [37]:
li = [0]
li[0] += 1
li

[1]

In [38]:
noised = [(10,20),(2,3)]
noised[-2]

(10, 20)

In [39]:
5 > 5

False

In [40]:
def noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations):
    if index == 0 and sent_idx == 0:
        if check_left_right_checkd[0] == 0 or check_left_right_checkd[0] ==1:
            if direction == 1:
                if shift_amount + position_right > len(offsets) - 1:
                    check_left_right_checkd[0] += 1

                    direction = -1
                    shift_amount = random.choice([1,2,3,4])
                    return noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations)
                else:
                    start = offsets[position_left + shift_amount][0]
                    end = offsets[position_right + shift_amount][1]
                    print("position_left: ", position_left)
                    print("position_right: ", position_right)
                    print("shift_amount: ", shift_amount)
                    print("offsets[position_left - shift_amount][0]: ", offsets[position_left - shift_amount][0])
                    return (start, end)
            else:
                if position_left - shift_amount >= 0:
                    print("check_left_right_checkd: ", check_left_right_checkd[0])
                    start = offsets[position_left - shift_amount][0]
                    end = offsets[position_right - shift_amount][1]
                    print("position_left: ", position_left)
                    print("position_right: ", position_right)
                    print("shift_amount: ", shift_amount)
                    print("offsets[position_left - shift_amount][0]: ", offsets[position_left - shift_amount][0])
                    return (start, end)
                else:
                    check_left_right_checkd[0] += 1
                    direction = 1
                    shift_amount = random.choice([1,2,3,4])
                    return noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations)
        else:
            if sent_idx < len(updated_word_offsets_2344) - 1:
                #shift boundary rightwards
                span_end_location = (position_right + shift_amount) % position_right
                next_sentence_offsets = updated_word_offsets_2344[sent_idx + 1]
                if span_end_location <= len(next_sentence_offsets) - 1:
                    #shif boundary leftwards
                    if position_left + shift_amount > len(offsets) - 1:

                        span_start_location = (position_left + shift_amount) %  position_left
                        start_offsets = updated_word_offsets_2344[sent_idx + 1]
                        start = start_offsets[span_start_location][0] 
                    else:
                        start = offsets[position_left + shift_amount][0]
                    end_offsets = updated_word_offsets_2344[sent_idx + 1]
                    end = end_offsets[span_end_location][1] 
                    return (start, end)
                else:
                    return -1
                    
            else:
                return -1
    else:
        if check_left_right_checkd[0] == 0 or check_left_right_checkd[0] ==1:
            if direction == 1:
                if shift_amount + position_right > len(offsets) - 1:
                    check_left_right_checkd[0] += 1

                    direction = -1
                    shift_amount = random.choice([1,2,3,4])
                    return noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations)
                else:
                    start = offsets[position_left + shift_amount][0]
                    end = offsets[position_right + shift_amount][1]
                    print("position_left: ", position_left)
                    print("position_right: ", position_right)
                    print("shift_amount: ", shift_amount)
                    print("offsets[position_left - shift_amount][0]: ", offsets[position_left - shift_amount][0])
                    return (start, end)
            else:
                end_of_previous_span = all_noisy_annotations[-1]
                if position_left - shift_amount >= 0 and position_right - shift_amount >=0:
                    # check if the starting of the shifted span does not overlap with the
                    # last offset of the previous span
                    if offsets[position_left - shift_amount][0] > end_of_previous_span[1]:
                        print("check_left_right_checkd: ", check_left_right_checkd[0])
                        start = offsets[position_left - shift_amount][0]
                        end = offsets[position_right - shift_amount][1]
                        print("position_left: ", position_left)
                        print("position_right: ", position_right)
                        print("shift_amount: ", shift_amount)
                        print("offsets[position_left - shift_amount][0]: ", offsets[position_left - shift_amount][0])
                        return (start, end)
                    else:
                        check_left_right_checkd[0] += 1
                        direction = 1
                        shift_amount = random.choice([1,2,3,4])
                        return noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations)
                else:

                    check_left_right_checkd[0] += 1
                    direction = 1
                    shift_amount = random.choice([1,2,3,4])
                    return noise_span(direction, shift_amount, position_right,position_left, offsets, index, check_left_right_checkd, sent_idx, updated_word_offsets_2344, all_noisy_annotations)
        else:
            if sent_idx < len(updated_word_offsets_2344) - 1:
                #shift boundary rightwards
                span_end_location = (position_right + shift_amount) % position_right
    
                next_sentence_offsets = updated_word_offsets_2344[sent_idx + 1]
                if span_end_location <= len(next_sentence_offsets) - 1:
                    #shif boundary leftwards
                    if position_left + shift_amount > len(offsets) - 1:

                        span_start_location = (position_left + shift_amount) %  position_left
                        start_offsets = updated_word_offsets_2344[sent_idx + 1]
                        start = start_offsets[span_start_location][0] 
                    else:
                        start = offsets[position_left + shift_amount][0]
                    end_offsets = updated_word_offsets_2344[sent_idx + 1]
                    end = end_offsets[span_end_location][1] 
                    return (start, end)
                else:
                    return -1
                    
            else:
                return -1

In [151]:
import random
import copy

def generate_noisy_spans(filtered_spans_min_4_char, updated_word_offsets_2344, sentences_length_of_abstract_indexes):
    """
    For each sentence, generates a noisy version of each span
    following the given rules.
    Returns: list of [original_span, noisy_span]
    """
    # , updated_word_offsets_2344, sentences_length_of_abstract_indexes
    all_noisy_annotations = []
    for sent_idx, all_spans_sentence in zip(sentence_indices_with_spans, filtered_spans_min_4_char):
        print("index: ", sent_idx, " and all spans: ", all_spans_sentence)
        print("offsets: ", updated_word_offsets_2344[sent_idx])
        print("sentence_boundary: ", sentences_length_of_abstract_indexes[sent_idx])
        noisy_spans_each_sentene = [0]
        offsets = updated_word_offsets_2344[sent_idx]
        
        for ind, span in enumerate(all_spans_sentence):
            shift_dir = random.choice([-1,1])  # -1 = left, 1 = right
            shift_amount = random.choice([1, 2, 3, 4])
            print("shift_dir: ", shift_dir)
            print("shift_amount: ", shift_amount)
            check_left_right_checkd = [0]
            
            end = span["endnew"]
            start = span["startnew"]
            print("=========================================")
            print("the end is: ", end)
            print("the start is: ", start)
            # Find the index where the second element == 109 
            position_right = next((i for i, tup in enumerate(offsets) if tup[1] == end), -1) 
            position_left  = next((i for i, tup in enumerate(offsets) if tup[0] == start), -1) 
            print("position_left: ", position_left)  # Output: 15
            print("position_right: ", position_right)
            span_offsets_noised = noise_span(shift_dir, shift_amount, position_right,position_left, offsets,ind, check_left_right_checkd, sent_idx, updated_word_offsets_2344,all_noisy_annotations)
            if span_offsets_noised == -1:
                span_offsets_noised = (start, end)
            all_noisy_annotations.append(span_offsets_noised)
            print("span_offsets_noised: ", span_offsets_noised)

    return all_noisy_annotations


In [164]:
noisy_annotations = generate_noisy_spans(filtered_spans_min_4_char, updated_word_offsets_935, sentences_length_of_abstract_indexes)


index:  0  and all spans:  [{'text': ['du', 'dialogue', 'abstrait'], 'startnew': 68, 'endnew': 88}]
offsets:  [(0, 10), (11, 14), (15, 22), (23, 25), (26, 32), (33, 42), (43, 46), (47, 49), (50, 55), (56, 58), (59, 67), (68, 70), (71, 79), (80, 88)]
sentence_boundary:  (0, 88)
shift_dir:  -1
shift_amount:  1
the end is:  88
the start is:  68
position_left:  11
position_right:  13
check_left_right_checkd:  0
position_left:  11
position_right:  13
shift_amount:  1
offsets[position_left - shift_amount][0]:  59
span_offsets_noised:  (59, 79)
index:  1  and all spans:  [{'text': ['pré-formés'], 'startnew': 116, 'endnew': 126}, {'text': ['du', 'langage', 'naturel,'], 'startnew': 188, 'endnew': 207}]
offsets:  [(90, 93), (94, 101), (102, 115), (116, 126), (127, 130), (131, 137), (138, 144), (145, 147), (148, 153), (154, 157), (158, 166), (167, 173), (174, 176), (177, 187), (188, 190), (191, 198), (199, 207), (208, 209), (210, 217), (218, 220), (221, 229), (230, 232), (233, 242), (243, 246), (

In [165]:
noisy_annotations

[(59, 79),
 (138, 144),
 (174, 190),
 (516, 528),
 (569, 584),
 (828, 837),
 (914, 919),
 (928, 946),
 (1128, 1145),
 (1192, 1201),
 (1326, 1338),
 (1455, 1460)]

In [154]:
filtered_spans_min_4_char

[[{'text': ['du', 'dialogue', 'abstrait'], 'startnew': 68, 'endnew': 88}],
 [{'text': ['pré-formés'], 'startnew': 116, 'endnew': 126},
  {'text': ['du', 'langage', 'naturel,'], 'startnew': 188, 'endnew': 207}],
 [{'text': ['du', 'langage'], 'startnew': 505, 'endnew': 515},
  {'text': ['la', 'préformation'], 'startnew': 590, 'endnew': 605}],
 [{'text': ['pré-formés'], 'startnew': 856, 'endnew': 866},
  {'text': ['langue:'], 'startnew': 884, 'endnew': 891},
  {'text': ['pré-formés', 'multilingues:'], 'startnew': 936, 'endnew': 960}],
 [{'text': ['synopses', 'abstraites'], 'startnew': 1099, 'endnew': 1118},
  {'text': ['agents'], 'startnew': 1202, 'endnew': 1208}],
 [{'text': ['la', 'technologie'], 'startnew': 1349, 'endnew': 1363}],
 [{'text': ['pré-formés'], 'startnew': 1437, 'endnew': 1447}]]

In [155]:
filtered_spans_min_4_char_list_of_tuples = []
for spans in filtered_spans_min_4_char:
    for span in spans:
        filtered_spans_min_4_char_list_of_tuples.append((span['startnew'], span['endnew']))

In [156]:
filtered_spans_min_4_char_list_of_tuples

[(68, 88),
 (116, 126),
 (188, 207),
 (505, 515),
 (590, 605),
 (856, 866),
 (884, 891),
 (936, 960),
 (1099, 1118),
 (1202, 1208),
 (1349, 1363),
 (1437, 1447)]

In [157]:
text_highlight =text935

In [166]:
# Preserve original text and offsets
text2344 = """Correction et validation de la dépendance syntaxique dans le Rhapsodie de la banque d’arbres française parlée

Cet article présente les méthodes, les résultats et la précision du processus d’annotation syntaxique de la Rhapsodie Treebank de français parlé. Le Rhapsodie Treebank est un corpus de 33 000 mots annoté pour la prosodie et la syntaxe, sous licence dans son intégralité sous Creative Commons. L’annotation syntaxique contient deux niveaux: un niveau macro-syntactique, contenant une segmentation en unités illocutionnaires (y compris des marqueurs de discours, des parenthèses...) et un niveau micro-syntactique comprenant des relations de dépendance et diverses structures paradigmatiques, appelées constructions de pile, ces dernières étant particulièrement fréquentes et diversifiées dans le langage parlé. Le processus d’annotation micro-syntactique, présenté dans cet article, comprend une préparation semi-automatique de la transcription, l’application d’un analyseur de dépendance syntaxique, le transcodage des résultats d’analyse au schéma d’annotation Rhapsodie, la correction manuelle par plusieurs annotateurs suivie d’un processus de validation, et enfin l’application de règles de cohérence qui vérifient les erreurs courantes. Les bons scores d’accord inter-annotateurs sont présentés et analysés plus en détail. L’article comprend également la liste des fonctions utilisées dans l’annotation de dépendance et pour la distinction des différentes constructions de pieux et présente les idées qui sous-tendent ces choix."""

#offsets = filtered_spans_min_4_char_list_of_tuples
offsets = noisy_annotations
# offsets = [(56, 70),
#  (138, 144),
#  (208, 220),
#  (529, 549),
#  (593, 609),
#  (936, 960),
#  (646, 652),
#  (856, 866),
#  (724, 728),
#  (741, 745),
#  (989, 1000),
#  (1242, 1251)]

# 1. Store annotations separately (preserves offsets)
annotations = [{"start": s, "end": e, "text": text935[s:e]} for s, e in offsets]

# 2. Function to generate highlighted HTML
def generate_html_highlight(text, annotations, color="yellow"):
    """
    Creates HTML with highlighted spans without modifying original text
    Preserves original offsets for future use
    """
    # Sort annotations by start position
    sorted_ann = sorted(annotations, key=lambda x: x["start"])
    
    html_parts = []
    last_end = 0
    
    for ann in sorted_ann:
        # Add text before highlight
        html_parts.append(text[last_end:ann["start"]])
        
        # Add highlighted text
        html_parts.append(f'<span style="background-color: {color};">{ann["text"]}</span>')
        
        last_end = ann["end"]
    
    # Add remaining text
    html_parts.append(text[last_end:])
    
    return "".join(html_parts)

# 3. Generate and save HTML
html_output = generate_html_highlight(text935, annotations, "green")

with open("highlighted_text_2344_final.html", "w", encoding="utf-8") as f:
    f.write(f"""<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Highlighted Text</title>
    <style>
        body {{ font-family: Arial, sans-serif; line-height: 1.6; }}
        span {{ border-radius: 3px; padding: 0 2px; }}
    </style>
</head>
<body>
    <pre>{html_output}</pre>
</body>
</html>""")

print("Highlighted HTML saved to highlighted_text.html")

Highlighted HTML saved to highlighted_text.html


In [66]:
text935[633 : 723]

'En outre, la grande majorité du travail accompli dans ce domaine a été axée sur l’anglais.'

### highlighting with ideal spans:


In [170]:
from IPython.display import HTML
HTML(html_output)

### highlighting with 3 different run of the code to generate the noisy spans for the same text

In [168]:
from IPython.display import HTML
HTML(html_output)

In [169]:
from IPython.display import HTML
HTML(html_output)

In [52]:
from IPython.display import HTML
HTML(html_output)